### DISCUSSION TITLE
코드의 결과를 설명해주세요

### DISCUSSION THREAD

**Discussion by publdaze** [publdaze/js-study#98] [this]
- 코드의 결과를 설명해주세요

```js
var value = 1;

var obj = {
  value: 100,
  foo: function() {
    console.log("foo's this: ",  this);
    console.log("foo's this.value: ",  this.value);
    function bar() {
      console.log("bar's this: ",  this);
      console.log("bar's this.value: ", this.value);
    }
    bar();
  }
};

obj.foo();
```

  - 중첩함수의 this 바인딩을 일치 시키려먼 어떻게 변경해야할까요?
    - call, apply, bind 메소드의 차이점은 무엇인가요?

**Comment by publdaze** [comment]
> * 코드의 결과를 설명해주세요
> 
> ```js
> var value = 1;
> 
> var obj = {
>   value: 100,
>   foo: function() {
>     console.log("foo's this: ",  this);
>     console.log("foo's this.value: ",  this.value);
>     function bar() {
>       console.log("bar's this: ",  this);
>       console.log("bar's this.value: ", this.value);
>     }
>     bar();
>   }
> };
> 
> obj.foo();
> ```

- foo 호출
- 호출 시점에 this 바인딩 결정 -> foo 메서드의 this에는 호출한 obj가 바인딩 됨
- 따라서
    "foo's this: {value: 100, foo: f}"
    "foo's this.value: 100"
- foo 메소드 내부에서 bar 호출
- 일반 함수로 호출 -> this에 전역 객체 바인딩
- 따라서
    "foo's this: Window {...}"
    "foo's this.value: 1"
    *여기서 1은 코드 1번 라인에 var로 선언된 value가 전역객체의 프로퍼티로 들어가기 때문

> * 중첩함수의 this 바인딩을 일치 시키려먼 어떻게 변경해야할까요?
- foo 내부에 변수를 선언해 foo 메소드의 this 값을 할당해주면 obj가 변수에 할당되고, 중첩 함수인 bar에서도 식별자 검색 시 상위 스코프인 foo의 식별자 참조가 가능하므로 변수에 할당된 값을 사용하면 동일한 this 바인딩 사용 가능
- 화살표 함수를 사용하는 경우, 화살표 함수는 this 바인딩을 하지 않음, 따라서 외부 함수의 this 바인딩을 그대로 사용하기 때문에 동일한 this 바인딩 사용 가능
- call, apply, bind 메소드로 바인딩할 객체를 명시적으로 넘겨주기
>   * call, apply, bind 메소드의 차이점은 무엇인가요?
- call - 바로 호출, 전달할 인수를 단일 원소로 제공
- apply - 바로 호출, 전달할 인수를 배열로 제공
- bind - 함수 반환

**Comment by pipisebastian** [comment]
> * 코드의 결과를 설명해주세요
> 
 ```js
var value = 1;

var obj = {
  value: 100,
  foo: function() {
	  // 객체의 메서드로써 호출되었기 때문에, obj가 바인딩됨
    console.log("foo's this: ",  this); // obj, {value : 100, foo : f}
    console.log("foo's this.value: ",  this.value); // 100
    function bar() {
	    // 일반함수로써 호출했기 때문에, 전역객체가 바인딩됨
      console.log("bar's this: ",  this); // 전역객체, window
      console.log("bar's this.value: ", this.value); // 1
    }
    bar();
  }
};

obj.foo();
```


> * 중첩함수의 this 바인딩을 일치 시키려먼 어떻게 변경해야할까요?

1. call/apply/bind로 명시적으로 this를 바인딩해줍니다.
2. 스코프체인을 이용하는 방식으로, 원하는 스코프, 예를 들어 상위 스코프의 this를 변수로 선언해줍니다. `_this`, `that`, `self` 등으로 변수명을 짓고, 하위 스코프에서 이를 참조하면 됩니다.
3. 화살표 함수를 사용할 수 있습니다(es6 이상) 화살표함수는 this를 binding하지 않아, 상위 스코프의 this를 바로 사용할 수 있습니다.  

>   * call, apply, bind 메소드의 차이점은 무엇인가요?
- 먼저 call, apply경우에는 첫번째 인자에 this로 만들고자하는 객체를 넣어주고, **바로 실행이** 됩니다.
    - `call` : 함수 인수를 쉼표로 구분하여 전달 
    - `apply`  : 함수 인수를 배열로 전달
- bind의 경우에는 this로 만들고자하는 객체를 넣어주면, this binding이 바뀐 함수를 리턴해줍니다. 그렇기때문에, 호출은 명시적으로 해줘야합니다.
